api/oauth.go

```diff
@@ -6,3 +6,18 @@ type OAuthToken struct {
 	RefreshToken string `jsonapi:"attr,refreshToken"`
 	ExpiresTs    int64  `jsonapi:"attr,expiresAt"`
 }
+
+// OAuthConfig is the API message for OAuthConfig.
+type OAuthConfig struct {
+	Endpoint      string `jsonapi:"attr,endpoint"`
+	ApplicationID string `jsonapi:"attr,applicationId"`
+	Secret        string `jsonapi:"attr,secret"`
+	RedirectURL   string `jsonapi:"attr,redirectUrl"`
+}
+
+// ExchangeToken is the API message for exchanging token.
+type ExchangeToken struct {
+	ID     int         `jsonapi:"attr,vcsId"`
+	Code   string      `jsonapi:"attr,code"`
+	Config OAuthConfig `jsonapi:"relation,config"`
```

Comment:
Request field is usually flattened. Without using relation.

Reponse field on the other hand uses relation for the composite type.

Reply:
Also if it's VCS specific, I would consider putting it under api/vcs.go

Another option is to use `jsonapi:"attr,id"`, it's the implementation job to interpret what's the ID for